File: lib/core/pairing_handler.dart
// ignore_for_file: non_constant_identifier_names
import 'package:unisync/utils/configs.dart';

import '../models/device_message/device_message.model.dart';
import 'device.dart';

enum PairState { unpaired, paired, pairRequested, markUnpaired, unknown }

abstract interface class PairOperation {
  void acceptPair();

  void rejectPair();

  void unpair();
}

class PairingHandler implements PairOperation {
  PairingHandler(this.device, {required this.onStateChanged});

  static const _method = (
    REQUEST: 'request',
    UNPAIR: 'unpair',
  );

  final Device device;
  final void Function(PairState state) onStateChanged;

  var _state = PairState.unknown;

  bool get isReady => _state != PairState.unknown;

  PairState get state => _state;

  Future<void> initiate() async {
    final result = await ConfigUtil.device.getPairedDeviceInfo(device.info.id);
    final info = result?.$1;
    final markUnpaired = result?.$2;
    if (result != null) {
      if (markUnpaired == true) {
        _state = PairState.markUnpaired;
      } else {
        _state = PairState.paired;
        if (info.toString() != device.info.toString()) {
          ConfigUtil.device.addPairedDevice(device.info);
        }
      }
    } else {
      _state = PairState.unpaired;
    }
    onStateChanged(_state);
  }

  @override
  void acceptPair() {
    if (_state == PairState.pairRequested) {
      device.sendMessage(DeviceMessage(
        time: DateTime.now().millisecondsSinceEpoch,
        type: DeviceMessage.Type.PAIR,
        header: DeviceMessageHeader(
          type: DeviceMessageHeader.Type.RESPONSE,
          status: DeviceMessageHeader.Status.SUCCESS,
          method: _method.REQUEST,
        ),
        body: {
          'accepted': true,
        },
      ));
      ConfigUtil.device.addPairedDevice(device.info);
      _state = PairState.paired;
      onStateChanged(_state);
    }
  }

  @override
  void rejectPair() {
    if (_state == PairState.pairRequested) {
      device.sendMessage(DeviceMessage(
        time: DateTime.now().millisecondsSinceEpoch,
        type: DeviceMessage.Type.PAIR,
        header: DeviceMessageHeader(
          type: DeviceMessageHeader.Type.RESPONSE,
          status: DeviceMessageHeader.Status.SUCCESS,
          method: _method.REQUEST,
        ),
        body: {
          'accepted': false,
        },
      ));
      _state = PairState.unpaired;
      onStateChanged(_state);
    }
  }

  @override
  void unpair() {
    if (_state == PairState.paired || _state == PairState.markUnpaired) {
      device.sendMessage(DeviceMessage(
        time: DateTime.now().millisecondsSinceEpoch,
        type: DeviceMessage.Type.PAIR,
        header: DeviceMessageHeader(
          type: DeviceMessageHeader.Type.NOTIFICATION,
          method: _method.UNPAIR,
        ),
      ));
      ConfigUtil.device.removePairedDevice(device.info);
      _state = PairState.unpaired;
      onStateChanged(_state);
    }
  }

  void handle(DeviceMessage message) {
    if (message.header.method == _method.REQUEST) {
      if (_state == PairState.unpaired) {
        _state = PairState.pairRequested;
        onStateChanged(_state);
      }
    }
    if (message.header.method == _method.UNPAIR) {
      ConfigUtil.device.removePairedDevice(device.info);
      _state = PairState.unpaired;
      onStateChanged(_state);
    }
  }
}

File: unisync_android/app/src/main/java/com/anhquan/unisync/core/PairingHandler.kt
package com.anhquan.unisync.core

import com.anhquan.unisync.models.DeviceMessage
import com.anhquan.unisync.utils.ConfigUtil

class PairingHandler(private val device: Device, private val onStateChanged: (PairState) -> Unit) {
    interface PairOperation {
        fun requestPair()

        fun unpair()
    }

    private object Method {
        const val REQUEST = "request"
        const val UNPAIR = "unpair"
    }

    val operation = object : PairOperation {
        override fun requestPair() {
            if (state == PairState.NOT_PAIRED) {
                device.sendMessage(
                    DeviceMessage(
                        type = DeviceMessage.Type.PAIR,
                        header = DeviceMessage.DeviceMessageHeader(
                            type = DeviceMessage.DeviceMessageHeader.Type.REQUEST,
                            method = Method.REQUEST
                        )
                    )
                )
                state = PairState.REQUESTED
                onStateChanged(state)
            }
        }

        override fun unpair() {
            if (state == PairState.PAIRED || state == PairState.MARK_UNPAIRED) {
                ConfigUtil.Device.removePairedDevice(device.info)
                device.sendMessage(
                    DeviceMessage(
                        type = DeviceMessage.Type.PAIR,
                        header = DeviceMessage.DeviceMessageHeader(
                            type = DeviceMessage.DeviceMessageHeader.Type.NOTIFICATION,
                            method = Method.UNPAIR
                        )
                    )
                )
                state = PairState.NOT_PAIRED
                onStateChanged(state)
            }
        }
    }

    init {
        ConfigUtil.Device.getDeviceInfo(device.info.id) { info, markUnpaired ->
            state = if (info != null) {
                if (markUnpaired == true) {
                    PairState.MARK_UNPAIRED
                } else {
                    if (info.toString() != device.info.toString()) {
                        ConfigUtil.Device.addPairedDevice(device.info)
                    }
                    PairState.PAIRED
                }
            } else {
                PairState.NOT_PAIRED
            }
            onStateChanged(state)
        }
    }

    enum class PairState {
        UNKNOWN, NOT_PAIRED, PAIRED, MARK_UNPAIRED, REQUESTED,
    }

    val isReady: Boolean get() = state != PairState.UNKNOWN

    var state: PairState = PairState.UNKNOWN
        private set

    fun handle(message: DeviceMessage) {
        when (message.header.method) {
            Method.REQUEST -> {
                if (message.body["accepted"] == true) {
                    ConfigUtil.Device.addPairedDevice(device.info)
                    state = PairState.PAIRED
                    onStateChanged(state)
                } else {
                    state = PairState.NOT_PAIRED
                    onStateChanged(state)
                }
            }
            Method.UNPAIR -> {
                ConfigUtil.Device.removePairedDevice(device.info)
                state = PairState.NOT_PAIRED
                onStateChanged(state)
            }
        }
    }
}